### PR TITLE
Mat gen

### PR DESCRIPTION
**Pull request by TeachRaccooon**
Reworked and applied matrix generation API. 
We now have a mat_gen_info class that controls all parameters used in matrix generation.
Moved matrix generation away from rl_util.hh to a new [rl_gen.hh](https://github.com/BallisticLA/RandLAPACK/blob/mat_gen/RandLAPACK/rl_gen.hh) file.

Question: @burlen i see that in a bunch of test routines we have `#include "RandLAPACK.hh"` plus a bunch of includes for individual computational routine files. Is there a readon why we keep the RandLAPACK.hh include statement? 

In the process of updating CQRRPT benchmarks to work with the new API, i realized that those are a mess and either need to be fully re-written or removed.
Since the CQRRPT project is nearly done, I went with removing those.
A couple of benchmark functions relying on RandLAPACK still remain so that we can test a project that uses RandLAPACK.

Additionally, a matrix generator for matrices with polynomially-decaying spectrum has been reworked per Riley's suggestion.
We can now control spectrum generation via both the condition number and an exponent.

### CODE REVIEW

File: RandLAPACK/drivers/rl_cqrrpt.hh [hunk 1]
```diff
@@ -365,7 +365,7 @@ int CQRRPT<T, RNG>::call(
     blas::syrk(Layout::ColMajor, Uplo::Upper, Op::Trans, k, m, 1.0, A_dat, m, 0.0, R_sp_dat, k);
     if(lapack::potrf(Uplo::Upper, k, R_sp_dat, k)){
         if(this->verbosity)
-            printf("CHOLESKY FACTORIZATION FAILED.\n");
+            throw std::runtime_error("Cholesky decomposition failed.");
```
Comment (rileyjmurray) [inline RandLAPACK/drivers/rl_cqrrpt.hh:368]: [❤️](https://emojipedia.org/red-heart/)

File: RandLAPACK/rl_gen.hh [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+}
+
+/// Generates matrix with the following singular values:
+/// sigma_i = 1 / (i + 1)^pow (first k * 0.2 sigmas = 1
+/// Output matrix is m by n of rank k.
+/// In later case, left and right singular vectors are randomly-generated
+/// and orthogonaized.
+/// Boolean parameter 'diag' signifies whether the matrix is to be
+/// generated as diagonal.
+/// Parameter 'cond' signfies the condition number of a generated matrix.
```
Comment (rileyjmurray) [inline RandLAPACK/rl_gen.hh:103]: Update me

File: RandLAPACK/rl_gen.hh [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+            A.resize(k * k);
+        }
+        lapack::lacpy(MatrixType::General, k, k, S.data(), k, A.data(), k);
+    } else {
+        RandLAPACK::gen::gen_singvec(m, n, A, k, S, state);
+    }
+}
+
+/// Generates matrix with the following singular values:
+/// sigma_i = e^((i + 1) * -pow) (first k * 0.2 sigmas = 1
```
Comment (rileyjmurray) [inline RandLAPACK/rl_gen.hh:151]: should be first 0.1 sigmas

**Review by rileyjmurray — COMMENTED** [review]
Looks good other than the documentation comments I left (and any issues you might identify while correcting those).

Merge at your discretion.

File: RandLAPACK/drivers/rl_cqrrpt.hh (hunk 1 above)
Comment (TeachRaccooon) [reply RandLAPACK/drivers/rl_cqrrpt.hh:368]: Baste.

File: RandLAPACK/rl_gen.hh (hunk 2 above)
Comment (TeachRaccooon) [reply RandLAPACK/rl_gen.hh:103]: You have now been updated.

File: RandLAPACK/rl_gen.hh (hunk 3 above)
Comment (TeachRaccooon) [reply RandLAPACK/rl_gen.hh:151]: Updated.

**Pull request merged**